vllm/v1/worker/gpu_model_runner.py

```diff
@@ -832,6 +900,44 @@ def make_sampling_metadata(
                 self.top_p_cpu_tensor[:self.num_reqs], non_blocking=True)
             self.top_k[:self.num_reqs].copy_(
                 self.top_k_cpu_tensor[:self.num_reqs], non_blocking=True)
+            if not self.no_reqs_with_penalties:
+                # Since syncing these tensors is expensive only copy them
+                # if necessary i.e. if there are requests which require
+                # penalties to be applied during sampling.
+                self.frequency_penalties[:self.num_reqs].copy_(
+                    self.frequency_penalties_cpu_tensor[:self.num_reqs],
+                    non_blocking=True)
+                self.presence_penalties[:self.num_reqs].copy_(
+                    self.presence_penalties_cpu_tensor[:self.num_reqs],
+                    non_blocking=True)
+                self.repetition_penalties[:self.num_reqs].copy_(
+                    self.repetition_penalties_cpu_tensor[:self.num_reqs],
+                    non_blocking=True)
+                # The prompt tokens are used only for applying penalties during
+                # the sampling process. Hence copy these tensors only when
+                # there are requests which need penalties to be applied.
+                self.prompt_tokens_tensor = \
+                    self._construct_prompt_tokens_tensor(
+                        requests, self.vocab_size, device=self.device)
```

Comment:
I am constructing this tensor only when skip_copy is False and not at every step. However I do construct this tensor using the prompt_token_id list. One alternative I can think of is to maintain the prompt tensors (on gpu) for each of these requests and pass the list of prompt tensors to the sampler. This will avoid converting each of the prompt token id list to a tensor every time there is a batch compaction. However in this case the sampler might need to construct a tensor from this list and then process it further. Is the second option what you are referring to here?

Reply:
OK I was thinking whether we can retain `prompt_token_ids` (just like other tensors in the input batch) rather than reconstructing it.
However, after second thoughts, I think we can keep the current version since the `prompt_token_ids` tensor can be very large and we will eventually manage other states (e.g., bin counts) instead of `prompt_token_ids` for penalties. I'm OK with the current implementation.